[FLINK-15790][k8s] Make FlinkKubeClient and its implementations asynchronous

wangyang0918 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

The `FlinkKubeClient` interface offers several methods which are synchronous (e.g. `FlinkKubeClient.createTaskManagerPod`, `FlinkKubeClient.stopPod`, `FlinkKubeClient.getPodsWithLabels`, etc). The problem is that these methods are directly used by the `KubernetesResourceManager` which calls them from the main thread. Since these methods perform I/O operations (sending and receiving network packages) they can potentially block the execution and hence should not happen from the `RpcEndpoint`'s main thread.

So the PR will make some interfaces the of FlinkKubeClient asynchronous, which potentially blocks the execution of RpcEndpoint's main thread. 

## Brief change log

* Make `createTaskManagerPod` and `stopPod` asynchronous


## Verifying this change
* This PR will **NOT** introduce any behavior change. All the current unit tests, e2e tests should work correctly.
* Manually test on a real K8s cluster for starting session, job submission, stop session, etc.


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

azagrebin commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java:
```diff
@@ -183,6 +183,14 @@
 		.withDescription("The user-specified annotations that are set to the TaskManager pod. The value could be " +
 			"in the form of a1:v1,a2:v2");
 
+	public static final ConfigOption<Integer> CLIENT_ASYNC_THREAD_POOL_SIZE =
```
Not sure whether we need to expose this, maybe thread pool with variable 1-4 threads would be good enough for all cases. If not then this option should be in the advanced category because user has to know details of the implementation to tune it.
Also, 2 threads may be enough by default.

Another idea, we discussed with @tillrohrmann, is to expose a general purpose executor for this kind of async background calls in RpcEndpoint handlers. Currently, we expose the Akka underlying dispatcher: `RpcEndpoint#getRpcService#getExecutor`. This is not nice because this executor cannot be really used. If we use it, we again risk to affect the `main thread`. We can change the implementation of `RpcEndpoint#getRpcService#getExecutor` to return another custom pool, like you create for the kube client now. As a first step, the executor could have just from 4 up to 16 threads, also without any config option. We could use that executor here as well. This is a separate issue which could help for other cases as well. Then we do not need this option `CLIENT_ASYNC_THREAD_POOL_SIZE`.

wangyang0918 replied:
Yes, this is exactly what i want too. Making `RpcEndpoint#getRpcService#getExecutor` return custom thread pool for I/O related execution is a very good idea. So i will remove this config option now and set the default thread pool size to 2.

wangyang0918 commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/ExecutorWrapper.java:
```diff
@@ -0,0 +1,67 @@
+			new ExecutorThreadFactory("FlinkKubeClient-IO"));
+	}
+
+	private ExecutorWrapper(Executor executor) {
```
The reason why i use `ExecutorWrapper` is to provide factory method to create a dedicated thread pool or reuse an existing one. However, i think this could also be done to make the argument of `Fabric8FlinkKubeClient` `Nullable`. I will go in this way.

azagrebin replied:
It does not have to be `@Nullable`.  It is better to create all dependencies outside of the class in its factory. The `Executors.newFixedThreadPool` can be created in `KubeClientFactory`.

azagrebin replied:
we can also shutdown any executor in `Fabric8FlinkKubeClient#close`, not sure why we need two fields `internalExecutorService` and `kubeClientExecutorService`

wangyang0918 replied:
In my opinion, who created the executor then it is responsible for its lifecycle. That's why i use the `internalExecutorService`. So if create the executor service outside of `Fabric8FlinkKubeClient` and shutdown it in `Fabric8FlinkKubeClient#close`. It may cause some problems since some other components are still using it.

Of cause, we could remove the `@Nullable` for `kubeClientExecutorService`, then we may need to add another argument to control whether need to shutdown it.
WDYT?

azagrebin replied:
You could write in comments that you hand over the lifecycle of the executor to the client. I do not see a big problem with that, it is created in its factory or tests, only for the client.
Alternatively, you could pass a factory for the `ExecutorService` to the client:
`Supplier<ExecutorService> asyncExecutorFactory`
then it is created and shutdown within the client.
It seems to me more clear approach with less ifs and vars.

Anyways, once we have the general purpose executor in RpcEndpoint [(FLINK-17024)](https://issues.apache.org/jira/browse/FLINK-17024), it can be refactored.

wangyang0918 replied:
Make sense. I will pass a executor service factory to `Fabric8FlinkKubeClient` and shutdown it in the `Fabric8FlinkKubeClient#close`.

wangyang0918 commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java:
```diff
@@ -61,158 +61,180 @@
-			throw new RuntimeException(
-				"Failed to find Deployment named " + clusterId + " in namespace " + this.nameSpace);
-		}
+		CompletableFuture.runAsync(() -> {
```
Good point. I will add a `retry` logic here. When all the retry times(3) is exhausted, we will throw the exception here.

azagrebin replied:
I think one of `stopPod` usages is not fixed in `KubernetesResourceManager#stopWorker`. The code to stop pod by its `ResourceID` can probably be deduplicated between KubernetesResourceManager#stopWorker` and KubernetesResourceManager#removePodIfTerminated`

azagrebin replied:
Side note: I do not quite understand the purpose of `FlinkKubeClient#handleException`.
Could we not just log it in `KubernetesClusterDescriptor` and `KubernetesResourceManager`?

wangyang0918 replied:
For `FlinkKubeClient#handleException`, i will remove it from `KubernetesResourceManager `. Since when we make `createTaskManagerPod` and `stopPod` asynchronous, it will not throw any exception then.

Actually, the `FlinkKubeClient#handleException` do not have meaningful operations. And it will make the exception stack confusing. I think it could be removed.

azagrebin replied:
I agree we can remove `FlinkKubeClient#handleException` as a follow-up

azagrebin replied:
One more question here if `FlinkKubeClient#createTaskManagerPod` fails due to the cluster long-time unavailability. We risk to enter into a frequent pod creation retry loop without any back-off. Is it handled somehow by fabric8 or okHttpClient client? Otherwise, we might want to follow-up on this.

azagrebin replied:
Also, it would be nice to add a test that KubernetesResourceManager retries creation of the pod if `FlinkKubeClient#createTaskManagerPod` future fails, using some testing implementation of `FlinkKubeClient`.

wangyang0918 replied:
This is a valid concern. I am afraid `fabric8` and `okHttpClient` could not cover this situation. Regarding to a follow-up, i tend to fix it in this PR.

I will move the retry logic into the `FlinkKubeClient#createTaskManagerPod`. In `KubernetesResourceManager kubeClient.createTaskManagerPod#whenComplete`, we just log the error. This could avoid the infinite retry. So in this situation, the slot allocation will timeout in `JobMaster#SlotPool`, and it will allocate again.

WDYT?

azagrebin replied:
I checked the JM implementation and it looks that JM decides when to timeout the pending request. Judging the existing active RM integrations (e.g. yarn), the RM should retry slot allocation until JM calls `RM#cancelSlotRequest` which happens in case of JM slot request timeout. Therefore, I still think that following `KubernetesResourceManager#onError` is more correct if internal fabric8 http request fails (still needs a unit test if not added yet).

My concern is more about a back-off strategy that the `KubernetesResourceManager` does not retry `FlinkKubeClient#createTaskManagerPod` too often but there is a certain time interval between retries. Maybe, something in fabric8 `Config` (like `KUBERNETES_REQUEST_TIMEOUT_SYSTEM_PROPERTY`) already can handle this?

wangyang0918 replied:
Thanks a lot for your comments. I will dig into fabric8 kubernetes client for more information. Also i will drop the current retry logics in `FlinkKubeClient#createTaskManagerPod` and still keep the retry in `KubernetesResourceManager`. A unit test should be added for such situation.

wangyang0918 replied:
After check the `fabric8`, `okhttp` client and some information about K8s apiserver, i believe that the internal `requestTimeout` and `connectionTimeout` could completely solve the "retry too often" problems. Since the connection may be normal and the apiserver return with some error code(`403, 404, 500`, etc.) immediately. 

So i think we still need to add a retry internal in `KubernetesResourceManager`. I have leave the default value to 3s. Also i add a test `testCreateTaskManagerPodFailedAndRetry` for creating taskmanager pod failed and retry.

azagrebin replied:
Alright, I think this should work in most cases. As a follow-up, we might want to make it configurable or add back-off (similar to 'NETWORK_REQUEST_BACKOFF_INITIAL'/`NETWORK_REQUEST_BACKOFF_MAX`) if this causes problems of retrying not fast enough in the beginning or too fast.

wangyang0918 replied:
Thanks for your suggestion. I think the initial and max back-off time could be a follow-up ticket.

zhengcanbin replied:
> Alright, I think this should work in most cases. As a follow-up, we might want to make it configurable or add back-off (similar to `NETWORK_REQUEST_BACKOFF_INITIAL`/`NETWORK_REQUEST_BACKOFF_MAX`) if this causes problems of retrying not fast enough in the beginning or too fast.

Thanks, @azagrebin ! Sorry to jump in, and this is a really good improvement, we have suffered such pains that the K8s API Server hangs after attacked by the infinite creation request from `KubernetesResourceManager`, the whole K8s cluster is unstable in such cases.

azagrebin replied:
Thanks for the feedback @zhengcanbin 
We might want to investigate your experience. Previously, we did not have the retry discussed in this PR. Therefore, there may be other sources of these infinite creation requests, e.g. from the #onError callback, depending how the k8s client works internally.
If there is already an issue for your problem (or you could create), could you link to https://issues.apache.org/jira/browse/FLINK-17127 or share your problem there? We might want to have a back-off also for the #onError callback depending on the actual problem.

zhengcanbin replied:
@azagrebin Thanks for the reply! I was planning to dive into this problem to find the root cause and come up with a solution for it. I will try to reproduce this kind of problem, then file an issue, if necessary, to track it and link it to [FLINK-17127](https://issues.apache.org/jira/browse/FLINK-17127), meanwhile, give you more feedback.

azagrebin replied:
Sounds good, thank you @zhengcanbin !

zhengcanbin replied:
Hi @azagrebin! I reproduce this problem and create two tickets to track some of the problems I recently found in the `KubernetesResourceManager`. Looking forward to your feedback.
[FLINK-17176](https://issues.apache.org/jira/browse/FLINK-17176)
[FLINK-17177](https://issues.apache.org/jira/browse/FLINK-17177)

wangyang0918 commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java:
```diff
@@ -61,158 +61,180 @@
+					if (address == null || address.isEmpty()) {
+						address = service.getStatus().getLoadBalancer().getIngress().get(0).getHostname();
+					}
+				} else {
+					address = this.internalClient.getMasterUrl().getHost();
+					endpointPort = getServiceNodePort(service, RestOptions.PORT);
 				}
-			} else {
-				address = this.internalClient.getMasterUrl().getHost();
-				restPort = getServiceNodePort(service, RestOptions.PORT);
+			} else if (service.getSpec().getExternalIPs() != null && service.getSpec().getExternalIPs().size() > 0) {
+				address = service.getSpec().getExternalIPs().get(0);
 			}
-		} else if (service.getSpec().getExternalIPs() != null && service.getSpec().getExternalIPs().size() > 0) {
-			address = service.getSpec().getExternalIPs().get(0);
-		}
-		if (address == null || address.isEmpty()) {
-			return null;
-		}
-		return new Endpoint(address, restPort);
+			if (address == null || address.isEmpty()) {
+				return Optional.empty();
+			}
+			return Optional.of(new Endpoint(address, endpointPort));
```
Make sense. I will update this part.

wangyang0918 commented:
@azagrebin Thanks for your review. I have addressed most of your comments. Since some others are not applicable when we only set `createTaskManagerPod` and `stopPod` as asynchronous.

Would you please take a look again at your convenience?

wangyang0918 commented:
@azagrebin Thanks for your review. I have addressed most of your comments. Since some others are not applicable when we only set `createTaskManagerPod` and `stopPod` as asynchronous.
Would you please take a look again at your convenience?

rmetzger commented:
Side note: I filed a ticket for the (unrelated) failure on Azure: https://issues.apache.org/jira/browse/FLINK-17006

azagrebin commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java:
```diff
@@ -61,158 +61,180 @@
+				} else {
+					address = this.internalClient.getMasterUrl().getHost();
+					endpointPort = getServiceNodePort(service, RestOptions.PORT);
 				}
```
what do you about introducing `getLoadBalancerRestEndpoint`?
imo, `getRestEndPointFromService` still looks quite big :)

wangyang0918 replied:
I will move them to a new function `getLoadBalancerRestEndpoint`.

azagrebin commented:
@wangyang0918 Thanks for addressing the comments.
I left couple of more questions in the unresolved inline comments to resolve them before merging.

wangyang0918 commented:
@azagrebin Thanks a lot for your second round review. I have integrated the unresolved comments. Could you take another look?

wangyang0918 commented:
@flinkbot run azure

azagrebin commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesTestBase.java:
```diff
@@ -84,7 +85,10 @@ public void setup() throws Exception {
+		flinkKubeClient = new Fabric8FlinkKubeClient(
+			flinkConfig,
+			kubeClient,
+			Executors::newDirectExecutorService);
```
`Fabric8FlinkKubeClientTest` uses a direct executor which technically allows not to call `get()` on the pod creation/stop futures in the corresponding tests. I would still call `get()` to avoid confusion in case of any future executor refactorings

wangyang0918 replied:
Thanks. I will remove the direct executor and use `get` explicitly.

wangyang0918 replied:
I think we need to keep the direct executor in `KubernetesResourceManagerTest`. So then once the `KubernetesResourceManager` calls `createTaskManagerPod/stopPod` internally, we could do the taskmanager check immediately.

azagrebin replied:
I did not mean to remove `Executors::newDirectExecutorService` in tests, only to call `get()` in tests of `Fabric8FlinkKubeClientTest` for newly return futures.

wangyang0918 replied:
Got it. I will rebase the master and update this part.

wangyang0918 commented:
@azagrebin I have integrated your improvements and rebased the new master branch. Also i add the retry logics to `createTaskManagerPod` and `stopPod`. When the operation still fails, the `KubernetesResourceManager` will just log the error. I add two tests for the new introduced retry logics, `testCreateFlinkTaskManagerRetryExhaustedAndFailed` and `testCreateFlinkTaskManagerRetryAndSucceed`.

wangyang0918 commented:
@flinkbot run azure

azagrebin commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesResourceManagerTest.java:
```diff
@@ -323,6 +330,18 @@ public void testGetCpuCoresNumSlots() {
 
+	@Test
+	public void testCreateTaskManagerPodFailedAndRetry() throws Exception {
+		flinkKubeClient = spy(createFlinkKubeClient());
```
According to Flink code style, we avoid using mockito mocks.
We usually create proper testing implementations of classes.
E.g. in this case, we could have:
```
private static FlinkKubeClient createTestingFlinkKubeClientAllocatingPodsAfter(
    int numberOfRetries,
    AtomicInteger retries) { ... }
```
which fails and increments `retries` until it reaches `numberOfRetries` and succeeds.

wangyang0918 replied:
Actually, i just want to reuse the `verify` with timeout in `mockito`. So i use the `spy`. But i think your suggestion makes sense, i will use the testing implementations instead.

azagrebin commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesResourceManagerTest.java:
```diff
@@ -398,4 +421,19 @@ private void terminatePod(Pod pod) {
+		return new Fabric8FlinkKubeClient(flinkConfig, kubeClient, () -> kubeClientExecutorService) {
+			@Override
+			public CompletableFuture<Void> createTaskManagerPod(KubernetesPod kubernetesPod) {
+				return CompletableFuture.runAsync(
+					() -> {
+						throw new RuntimeException("Exception");
+					},
+					kubeClientExecutorService
+				);
```
```suggestion
				return CompletableFuture.completeExceptionally(new RuntimeException("Exception"));
```
This may be good enough for the test.

wangyang0918 replied:
I am not what do you mean here. Since the `createTaskManagerPod` needs to return a `CompletableFuture<Void>`.

azagrebin replied:
Sorry, I meant you can return `FutureUtils.completedExceptionally(new RuntimeException("Exception"))`. This will create already completed with an Exception future.

wangyang0918 replied:
Nice suggestion.

azagrebin commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesResourceManagerTest.java:
```diff
@@ -323,6 +330,18 @@ public void testGetCpuCoresNumSlots() {
+	@Test
+	public void testCreateTaskManagerPodFailedAndRetry() throws Exception {
+		flinkKubeClient = spy(createFlinkKubeClient());
+		TestingKubernetesResourceManager testRM = createAndStartResourceManager(flinkConfig);
```
Not sure, how this would work, `KubernetesResourceManager` creates its own `kubeClient` internally, different from `flinkKubeClient`. I would suggest to inject `kubeClient` into the constructor of `KubernetesResourceManager` in `KubernetesResourceManagerFactory`.

wangyang0918 replied:
This could work because we override the `createFlinkKubeClient` in `TestingKubernetesResourceManager`. Injecting the `kubeClient` into the constructor of `KubernetesResourceManager` also makes sense to me. I will go in this way.

wangyang0918 commented:
@azagrebin I have integrated the recent comments. Please take a look at your convenience.

azagrebin commented:
Thanks for addressing the comments @wangyang0918 

I also changed a bit test to wait for the actual pod creation with the latch and reduced the retry interval for tests to keep test run time low.

One hotfix commit introduces `KubernetesResourceManagerConfiguration`. It is created by factory and  injected into the `KubernetesResourceManager`. Ideally, there should be no logic in constructor but in factories. The object should get only needed dependencies to simplify mocking them in tests.

If you are ok with my additions, the PR looks good to me.

wangyang0918 commented:
@flinkbot run travis
@flinkbot run azure

wangyang0918 commented:
@azagrebin Thanks for the following fix. I am completely ok with the additional changes. Moreover, i am not sure whether i could squash your commits. So i still leave them there until you merge this PR.

azagrebin commented:
@wangyang0918 
I would suggest one more follow-up.
We call `FlinkKubeClient#stopAndCleanupCluster` in `KubernetesResourceManager#internalDeregisterApplication` which is an abstract method of `ResourceManager` called from `deregisterApplication`. The `deregisterApplication` is async. We could easily make the abstract method `ResourceManager#internalDeregisterApplication` also async and return its result future from `deregisterApplication`, also with logging in case of error. All other implementations does not have to be async, they could simply return `CompletableFuture.completedFuture(Acknowledge.get())` for now.

wangyang0918 commented:
@azagrebin Thanks for the additional comments. First, i agree with you after these changes, `FlinkKubeClient#stopAndCleanupCluster` could also be async.

However, i am just thinking what could we benefit from this change? One advantage in my mind is we could make all the `internalDeregisterApplication` implementations executed asynchronously in the `#getRpcService#getExecutor()`. And then Flink cluster could shut down a little faster. 

Do you have other thoughts or concerns?

azagrebin commented:
@wangyang0918 
True, this could also help in other integrations. We can consider them separately. I do not know atm whether each `internalDeregisterApplication` implementation is indeed blocking.

wangyang0918 commented:
@azagrebin For YARN, the `AMRMAsyncClient.unregisterApplicationMaster` is a blocking call. Currently, the same for the `FlinkKubeClient#stopAndCleanupCluster`. I am not very familiar with mesos, but from the `SchedulerDriver` implementation, i do not see it is an async call.

So it seems all the `internalDeregisterApplication` implementations now are blocking. I prefer to create a new ticket "Make ResourceManager#internalDeregisterApplication and implementation asynchronous", which will deal with them together.

azagrebin commented:
@wangyang0918 I agree to move this to a follow-up ticket

wangyang0918 commented:
@azagrebin A new follow-up tickets has been created.
https://issues.apache.org/jira/browse/FLINK-17154

wangyang0918 commented:
@flinkbot run azure

wangyang0918 commented:
@azagrebin Since a new K8s related PR has been merge, i rebase the latest master. I update the PR with the following changes to resolve conflicts.
* Update `Fabric8FlinkKubeClient#getRestEndpoint`
* Update `Fabric8FlinkKubeClientTest`

Could you please have a final check and help with the merging?

wangyang0918 commented:
Hi @azagrebin, both the travis and azure has passed. Could you help with merging this PR? Also for the backport #11733. I really hope it could catch up with the upcoming release-1.10.1.
